### ISSUE TITLE
UPNP add_port_mapping() returns error 26 even though discover() returns 0 beforehand

### ISSUE THREAD

**Issue by Anotic** [godotengine/godot#70458]
### Godot version

4.0 beta 9

### System information

Windows 10, Ryzen 5 3600, RTX 2060. Godot 4 beta 9 Forward+

### Issue description

When I use the code on the latest version of the Godot Docs the upnp.discover() function returns 0 saying that it was successfull.
But upnp.add_port_mapping() returns 26 stating that I need to use upnp.discover() first. Even though I already have and it was successful.

### Steps to reproduce

I simply ran this code from the Godot Latest Docs and tried to find what was going wrong by just printing it.

func _upnp_setup(port):
	var upnp = UPNP.new()
	var err = upnp.discover()
	
	print(err) #Prints 0 for me showing it was a success
	
	if err != OK:
		push_error(str(err))
		emit_signal("upnp_completed", err)
		print(err)
		return
	
	print(upnp.get_gateway()) #Prints 26 saying I need to run discover first
	
	if upnp.get_gateway() and upnp.get_gateway().is_valid_gateway():
		upnp.add_port_mapping(port, port, ProjectSettings.get_setting("application/config/name"), "UDP")
		upnp.add_port_mapping(port, port, ProjectSettings.get_setting("application/config/name"), "TCP")
		emit_signal("upnp_completed", OK)
		print(OK)
		upnp.delete_port_mapping(port)

### Minimal reproduction project

N/A

**Comment by Calinou** [comment]
Can you reproduce this on 3.5.1?

**Comment by Citonaa** [comment]
using the same code in 3.5.1 retuns 0 for upnp.discover() and returns [Object:null] for upnp.get_gateway()

**Comment by hcoura** [comment]
I took a look on this one, I don't think this is a regression, nor a bug necessarily for that matter.

`upnp.discover()` may return `0` (success) if it found any device even if their igd_status is invalid
thus, `upnp.get_gateway` will return null and `upnp.add_port_mapping` will return `26` and display this [message](https://github.com/godotengine/godot/blob/master/modules/upnp/doc_classes/UPNP.xml#L241):

```No gateway available. You may need to call [method discover] first, or discovery didn't detect any valid IGDs (InternetGatewayDevices).```

suggesting one may need run `upnp.discover()`...


There is no significant change between 3.5.1 and 4.x on the related source files and `using the same code in 3.5.1 retuns 0 for upnp.discover() and returns [Object:null] for upnp.get_gateway()` pretty much says the behaviour is the same there

@Citonaa I think this line in your steps to reproduce is not correct `print(upnp.get_gateway()) #Prints 26 saying I need to run discover first` this function should return either a device or null in either [4.x](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L267) or [3.5.1](https://github.com/godotengine/godot/blob/3.5.1-stable/modules/upnp/upnp.cpp#L268)


------

I guess the only question is if this situation should be made more clear and if `discover` should return success in the case of devices found that fail `is_valid_gateway()` check.

**Comment by Citonaa** [comment]
Tried again on latest Godot 4 and print(upnp.get_gateway()) does return null. Not sure what was happening before as I was getting 0 and then 26 printed one after the other when I made the issue and assumed that print(upnp.get_gateway()) was what printed it./

**Comment by ywmaa** [comment]
I have this problem too in Godot 4.0.1.

**Comment by mhilbrunner** [comment]
@ywmaa Could you post your router model (and possibly firmware version)?

**Comment by ywmaa** [comment]
@mhilbrunner 
yeah I am very sure that upnp option is enabled in the router, and just in case I disabled firewall in the router.


the router is huawei DG8045

I don't know how to get firmware version.

**Comment by mhilbrunner** [comment]
*Software Version* sounds right, yeah.

There seem to be [various](https://www.snbforums.com/threads/miniupnpc-does-not-support-huawei-e3276-4g-modem.32334/) [reports](https://miniupnp.tuxfamily.org/forum/viewtopic.php?t=1693) around the web of people having [trouble](https://discuss.uavmatrix.com/d/5141-Have-trouble-when-try-to-forward-the-port-using-Huawei-E3276) with Huawei routers in general and Huawei routers and MiniUPnP specifically.

We even got a report for Godot in 2019 [here](https://github.com/godotengine/godot/issues/25008#issuecomment-552170589) which seems similar, so maybe this is a specific problem with (some?) Huawei routers and/or MiniUPnPc.

As we don't maintain MiniUPnP, this problem sees unlikely to be fixable on our side.

**Comment by ywmaa** [comment]
This is solved for me by manually port forwarding in the router, now upnp works, and gets the router, also can fetch the global IP,  but I think this makes it useless for my router, after all I had to port forward myself, but my router is from huawei which can be the cause of the problem.

**Comment by Anotic** [comment]
> This is solved for me by manually port forwarding in the router, now upnp works, and gets the router, also can fetch the global IP, but I think this makes it useless for my router, after all I had to port forward myself, but my router is from huawei which can be the cause of the problem.

i can manually port suff but the whole point of upnp is a semi automated way of doing it :/

**Comment by Calinou** [comment]
In general, UPnP in 2023 is a lot less worthwhile than it was back in 2010-2012. Something like https://github.com/godotengine/godot-proposals/issues/434 should be implemented to replace it entirely at some point.

**Comment by Anotic** [comment]
using a NAT server in godot would be great

**Comment by chrom007** [comment]
I have same problem with TPLink Archer C6 v2 :cry:

**Comment by FeralBytes** [comment]
May I request that we update miniUPnP to the latest version 2.2.4. I am also having issues with a TP Link Archer AX20 router, but it works fine in [Transmission](http://www.transmissionbt.com/) and seems that there was an issue fixed late last year for TP Link routers that may fix some of the UPnP issues. I also see that we are currently using version 2.2.3 last updated on [19NOV2021](https://github.com/godotengine/godot/commit/432ac521ef50ca41a81cffc1770cb46f72a422f6). Thank you all for the amazing game engine that is Godot.

**Comment by FeralBytes** [comment]
@akien-mga thanks for pointing that out, sorry I missed it. Is it possible then that we are running into the same statusinfo() exception that I was encountering in Python, and perhaps that is why we are not able to use the IGD correctly in Godot?

**Comment by mhilbrunner** [comment]
Thanks for that testing, I'll investigate it (althought thats a bit hard without a TP device for testing) - at least its a direction to look into.

**Comment by FeralBytes** [comment]
@mhilbrunner is there any thing that I can do to test a fix, or do you know what code I could adjust to test my theory about the issue of calling statusinfo()?

**Comment by mhilbrunner** [comment]
@FeralBytes The code for UPnP lives in [upnp.cpp](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp) and [upnp_device.cpp](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp_device.cpp).

If I understand it correctly, the Python `upnpc.statusinfo()` function is a wrapper around [this](https://github.com/miniupnp/miniupnp/blob/e663006e68628fff4028cbed040759cf51011b92/miniupnpc/src/miniupnpcmodule.c#L246) C function in the MiniUPnP client, which is basically a call to [UPNP_GetStatusInfo](https://github.com/miniupnp/miniupnp/blob/e663006e68628fff4028cbed040759cf51011b92/miniupnpc/src/upnpcommands.c#L122) plus some setup/teardown. If I'm not missing anything, we do not seem to call `UPNP_GetStatusInfo` in any of our code.

Now, what we _do_ do is after finding devices in [UPNP.discover()](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L46), we [add them to our device list](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L93) and then call [parse_igd()](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L109) on their description URL (which in turn calls [load_description()](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L105), which is just a small wrapper using `miniwget` to download the XML from that URL).

If I were to hazard a guess, this fails for some reason (maybe for the same reason that `statusinfo()` does, as it probably does something similar) - maybe the router doesn't return the XML as expected, or does not correctly identify itself as an IGD (InternetGatewayDevice), or something else is going on during that HTTP request.

Now as said, this is kinda hard to debug without having such a router around to actually look at the traffic. What _could_ be potentially interesting is if you could insert a debug print statement in [load_description](https://github.com/godotengine/godot/blob/master/modules/upnp/upnp.cpp#L105) and log the status code, the result body/XML and such and see what is actually received or if there are errors during `UPNP.discover()`. Maybe you should also try to set the `device_filter` parameter of `UPNP.discover()` to something other than `InternetGatewayDevice` (maybe empty string).

If you can find out the URL to your routers description URL for UPNP, building with [this PR](https://github.com/godotengine/godot/pull/68276) and manually setting the description URL to see if that works could also be interesting to skip the whole discovery process and pinpoint where the issue occurs.

If you are new to modifying the Godot source and building it, the documentation to get setup and started is [around here](https://docs.godotengine.org/en/stable/contributing/development/compiling/index.html), but I'm also happy to help you debug this if you want to join [contributors chat](https://chat.godotengine.org/) and ping me on the #networking channel to talk!
(Responses may take some time though :))

**Comment by drautb** [comment]
FWIW, I also have a TP-link router and have faced similar issues. I tracked it down to the fact that TP-link routers don't correctly implement the `GetStatusInfo` action ([bug report][1]), despite the router-provided XML declaring that it _is_ supported. When miniupnp tries to invoke that action, it receives an HTTP 500 response, and treats the device as disconnected.

[1]: https://community.tp-link.com/us/home/forum/topic/577840
